Doc/c-api/typeobj.rst

```diff
@@ -676,32 +676,51 @@ and :c:data:`PyType_Type` effectively act as defaults.)
-   memory buffers owned by the instance (using the freeing function corresponding
-   to the allocation function used to allocate the buffer), and call the type's
-   :c:member:`~PyTypeObject.tp_free` function.  If the type is not subtypable
-   (doesn't have the :c:macro:`Py_TPFLAGS_BASETYPE` flag bit set), it is
-   permissible to call the object deallocator directly instead of via
-   :c:member:`~PyTypeObject.tp_free`.  The object deallocator should be the one used to allocate the
-   instance; this is normally :c:func:`PyObject_Free` if the instance was allocated
-   using :c:macro:`PyObject_New` or :c:macro:`PyObject_NewVar`, or
-   :c:func:`PyObject_GC_Del` if the instance was allocated using
-   :c:macro:`PyObject_GC_New` or :c:macro:`PyObject_GC_NewVar`.
-
-   If the type supports garbage collection (has the :c:macro:`Py_TPFLAGS_HAVE_GC`
-   flag bit set), the destructor should call :c:func:`PyObject_GC_UnTrack`
+   The destructor function should remove all references which the instance owns
+   (e.g., call :c:func:`Py_CLEAR`), free all memory buffers owned by the
+   instance, and call the type's :c:member:`~PyTypeObject.tp_free` function to
+   free the object itself.
+
+   No guarantees are made about when an object is destroyed, except:
+
+   * Python will destroy an object immediately or some time after the final
+     reference to the object is deleted, unless its finalizer
+     (:c:member:`~PyTypeObject.tp_finalize`) subsequently resurrects the
+     object.
+   * An object will not be destroyed while it is being automatically finalized
+     (:c:member:`~PyTypeObject.tp_finalize`) or automatically cleared
+     (:c:member:`~PyTypeObject.tp_clear`).
+
+   .. warning::
+
+      The :c:member:`~PyTypeObject.tp_dealloc` function can be called from any
+      thread, although the :term:`GIL` will be held.
+
+   Python currently destroys an object immediately from :c:func:`Py_DECREF`
+   when the new reference count is zero, but this may change in a future
+   version.
+
+   It is recommended to call :c:func:`PyObject_CallFinalizerFromDealloc` at the
+   beginning of :c:member:`!tp_dealloc` to guarantee that the object is always
+   finalized before destruction.
+
+   If the type supports garbage collection (the :c:macro:`Py_TPFLAGS_HAVE_GC`
+   flag is set), the destructor should call :c:func:`PyObject_GC_UnTrack`
    before clearing any member fields.
 
-   .. code-block:: c
+   It is permissible to call :c:member:`~PyTypeObject.tp_clear` from
+   :c:member:`!tp_dealloc` to reduce code duplication and to guarantee that the
+   object is always cleared before destruction.  Beware that
+   :c:member:`!tp_clear` might have already been called.
+
+   If the type is heap allocated (:c:macro:`Py_TPFLAGS_HEAPTYPE`), the
+   deallocator should release the owned reference to its type object (via
+   :c:func:`Py_DECREF`) after calling the type deallocator.  See the example
+   code below.
```

Comment:
Doctest failures: we still need the `.. code-block:: c`, or because we already have C as the default for the document (set at the top), add `::` at the end of this paragraph.